### PR TITLE
2p revamp

### PR DESCRIPTION
**Pull request by HarryWangATX**

### CODE REVIEW

File: content/3_Silver/2P.mdx [hunk 1]
```diff
@@ -31,6 +31,133 @@ frequency: 2
...
 </Resources>
 
+<FocusProblem problem="book" />
+
+### Solution - Books
+
+We want to find the longest continuous segment of books that can be read within $t$ minutes. 
+
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.
+
```
Comment (frodakcin) [inline content/3_Silver/2P.mdx:41]: Think it would help to add another line saying Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segments $\texttt{left} \dots \texttt{right}-1$, we can use two-pointer (or, alternatively, binary search)

**Review by frodakcin — CHANGES_REQUESTED** [review]
Looks good. Just smth I think would be nice to add

**Review by darren-yao — APPROVED** [review]
The content looks good to me (I didn't read the code, but I trust that your code works and is of good quality). 

We probably want Python code eventually, but this isn't particularly urgent and shouldn't hold up pushing the module to live. 

Approved.

File: content/3_Silver/2P.mdx [hunk 2]
```diff
@@ -31,6 +31,139 @@ frequency: 2
 	</Resource>
 </Resources>
 
+## Two Pointers
+
+The two pointers method iterates two pointers across an array, to track the start and end
+of an interval, or two values in a sorted array that we are currently checking. Both pointers
+are monotonic; meaning each pointer starts at one end of the array and only moves in one
```
Comment (andrewwangva) [inline content/3_Silver/2P.mdx:38]: don't think monotonic is the right word here

File: content/3_Silver/2P.mdx [hunk 3]
```diff
@@ -31,6 +31,139 @@ frequency: 2
...
+
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.
+
+Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segments $\texttt{left} \dots \texttt{right}-1$, we can use two-pointer (or, alternatively, [binary search](/silver/binary-search)).
+
+For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
+
+We will keep track of the total time with $\texttt{cur}$.
+
+Once $\texttt{cur}$ exceeds $t$, we remove the current $\texttt{right}$-indexed book and decrement $\texttt{right}$ by 1.
```
Comment (andrewwangva) [inline content/3_Silver/2P.mdx:55]: seems kinda extra wordy in the last two sentences. Both lines are kinda redundant so just remove both sentences.

File: content/3_Silver/2P.mdx [hunk 4]
```diff
@@ -31,6 +31,139 @@ frequency: 2
...
+
+We will keep track of the max segment size with $\texttt{ans}$, and take the maximum between $\texttt{ans}$ and $\texttt{right} - \texttt{left}$.
```
Comment (andrewwangva) [inline content/3_Silver/2P.mdx:57]: also kind of self-explanatory that we are trying to find max.

Comment (andrewwangva) [reply content/3_Silver/2P.mdx:57]: maybe it's better to reiterate pointers move in one direction and the right pointer simply keeps moving to the right instead of starting over?

File: content/3_Silver/2P.mdx [hunk 5]
```diff
@@ -31,6 +31,144 @@ frequency: 2
 	</Resource>
 </Resources>
 
+## Two Pointers
+
+The two pointers method iterates two pointers across an array, to track the start and end
+of an interval, or two values in a sorted array that we are currently checking. 
+Typically, each pointer starts at seperate ends and moves strictly in one direction. 
+However, as shown in the focus problem below, the two pointers do not have to start at seperate ends.
```
Comment (thecodingwizard) [inline content/3_Silver/2P.mdx:39]: oops I found this rather confusing (no idea how to fix though) D:

File: content/3_Silver/2P.mdx [hunk 6]
```diff
@@ -31,6 +31,144 @@ frequency: 2
...
+
+<FocusProblem problem="book" />
+
+## Solution - Books
+
+We want to find the longest continuous segment of books that can be read within $t$ minutes. 
+
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.
+
+Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segments $\texttt{left} \dots \texttt{right}-1$, we can use two-pointer (or, alternatively, [binary search](/silver/binary-search)).
```
Comment (thecodingwizard) [inline content/3_Silver/2P.mdx:50]:
```suggestion
Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segment $\texttt{left} \dots \texttt{right}-1$, we can use two-pointers (or, alternatively, [binary search](/silver/binary-search)).
```

File: content/3_Silver/2P.mdx [hunk 7]
```diff
@@ -31,6 +31,144 @@ frequency: 2
...
+
+For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
+
+We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
+
+We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$ and increment $\texttt{left}$ by 1.
```
Comment (thecodingwizard) [inline content/3_Silver/2P.mdx:56]: we never defined cur. maybe write this out explicitly?

File: content/3_Silver/2P.mdx (hunk 5 above)
Comment (HarryWangATX) [reply content/3_Silver/2P.mdx:39]: which part is confusing?

File: content/3_Silver/2P.mdx [hunk 8]
```diff
@@ -31,6 +31,144 @@ frequency: 2
...
+					break;
+				}
+			}
+
+			ans = Math.max(ans, right - left);
+			cur -= books[left++];
+		}
+
+		io.println(ans);
+		io.close();
```
Comment (SansPapyrus683) [inline content/3_Silver/2P.mdx:161]: 139 and 141 seem unnecessary

File: content/3_Silver/2P.mdx [hunk 9]
```diff
@@ -31,6 +31,144 @@ frequency: 2
...
+		int t = io.nextInt();
+
+		int[] books = new int[n];
+
+		for(int i = 0; i < n; i++) {
+			books[i] = io.nextInt();
+		}
+
+		int left = 0, right = 0, cur = 0, ans = 0;
+		while(left < n && right < n) {
```
Comment (SansPapyrus683) [inline content/3_Silver/2P.mdx:147]:
```suggestion
		while (left < n && right < n) {
```

File: content/3_Silver/2P.mdx [hunk 10]
```diff
@@ -31,6 +31,144 @@ frequency: 2
...
+			while(right < n) {
```
Comment (SansPapyrus683) [inline content/3_Silver/2P.mdx:148]:
```suggestion
			while (right < n) {
```

File: content/3_Silver/2P.mdx [hunk 11]
```diff
@@ -31,6 +31,144 @@ frequency: 2
...
+				cur += books[right++];
+				if(cur > t) {
```
Comment (SansPapyrus683) [inline content/3_Silver/2P.mdx:150]: if (cur > t) {

**Review by SansPapyrus683 — COMMENTED** [review]
my 2 cents

File: content/3_Silver/2P.mdx [hunk 12]
```diff
@@ -31,6 +31,144 @@ frequency: 2
...
+		Kattio io = new Kattio();
+
+		int n = io.nextInt();
+		int t = io.nextInt();
+
+		int[] books = new int[n];
+
+		for(int i = 0; i < n; i++) {
+			books[i] = io.nextInt();
+		}
```
Comment (HarryWangATX) [inline content/3_Silver/2P.mdx:144]:
```suggestion
		int n = io.nextInt();
		int t = io.nextInt();
		int[] books = new int[n];
		for(int i = 0; i < n; i++) {
			books[i] = io.nextInt();
		}
```

File: content/3_Silver/2P.mdx [hunk 13]
```diff
@@ -31,6 +31,142 @@ frequency: 2
...
+	int n, t, ans = 0;
+	cin >> n >> t;
+	vi books(n);
+	for(int i = 0; i < n; i++){
+		cin >> books[i];
+	}
+
+	int left = 0, right = 0, cur = 0;
+	while(left < n && right < n) {
+		while(right < n) {
```
Comment (SansPapyrus683) [inline content/3_Silver/2P.mdx:109]: oh didn't see these two while loops, could you add spaces for these

File: content/3_Silver/2P.mdx [hunk 14]
```diff
@@ -31,6 +31,142 @@ frequency: 2
...
+import java.io.*;
+
+public class Books {
+	public static void main(String[] args) {
+		Kattio io = new Kattio();
+
+		int n = io.nextInt();
+		int t = io.nextInt();
+		int[] books = new int[n];
+		for(int i = 0; i < n; i++) {
```
Comment (SansPapyrus683) [inline content/3_Silver/2P.mdx:140]: and this for loop lmao
```suggestion
		for (int i = 0; i < n; i++) {
```

**Review by SansPapyrus683 — APPROVED** [review]
pog

File: content/3_Silver/2P.mdx [hunk 15]
```diff
@@ -31,6 +31,139 @@ frequency: 2
...
+	cin.tie(0)->sync_with_stdio(0); // see /general/fast-io
+	if (sz(name)) {
+		freopen((name+".in").c_str(), "r", stdin); // see /general/io
+		freopen((name+".out").c_str(), "w", stdout);
+	}
+}
+
+
+int main() {
+	setIO();
```
Comment (nwatx) [inline content/3_Silver/2P.mdx:96]: since this is the focus problem, I think it would be beneficial to comment the code.

File: content/3_Silver/2P.mdx [hunk 16]
```diff
@@ -31,6 +31,139 @@ frequency: 2
...
+
+<FocusProblem problem="book" />
+
+## Solution - Books
+
+We want to find the longest continuous segment of books that can be read within $t$ minutes. 
+
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.
+
+Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segment $\texttt{left} \dots \texttt{right}-1$, we can use two-pointers (or, alternatively, [binary search](/silver/binary-search)).
```
Comment (nwatx) [inline content/3_Silver/2P.mdx:47]: ok maybe it's just me, but the binary search part might distract from the focus of the problem.

File: content/3_Silver/2P.mdx [hunk 17]
```diff
@@ -31,6 +31,139 @@ frequency: 2
...
+
+We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
+
+We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$, the total time of current segment, and increment $\texttt{left}$ by 1.
+
+**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right.**
+
+<Info title="Pro-tip">
+
+You can visualize these pointers as maintaing a sliding window of books for this problem.
```
Comment (nwatx) [inline content/3_Silver/2P.mdx:59]:
```suggestion
You can visualize these pointers as maintaining a sliding window of books for this problem.
```

**Review by nwatx — CHANGES_REQUESTED** [review]

File: content/3_Silver/2P.mdx [hunk 18]
```diff
@@ -31,6 +31,143 @@ frequency: 2
...
+
+Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segment $\texttt{left} \dots \texttt{right}-1$, we can use two-pointers (or, alternatively, [binary search](/silver/binary-search)).
+
+For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
+
+We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
+
+We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$, the total time of current segment, and increment $\texttt{left}$ by 1.
+
+**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right.**
```
Comment (dongliu0426) [inline content/3_Silver/2P.mdx:55]: > **The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right.**

add that because of this, the time complexity is O(N)

**Review by dongliu0426 — CHANGES_REQUESTED** [review]

File: content/3_Silver/2P.mdx (hunk 18 above)
Comment (dongliu0426) [reply content/3_Silver/2P.mdx:55]: not addressed? also, it might be better to say that the time complexity is O(N) because the left and right pointers move at most N steps in total.

Comment (HarryWangATX) [reply content/3_Silver/2P.mdx:55]: i moved the time comp down cus i thought i was a little more consistent (?)

Comment (HarryWangATX) [reply content/3_Silver/2P.mdx:55]: ill just move it up in the explanation ig

**Review by dongliu0426 — APPROVED** [review]

File: content/3_Silver/2P.mdx [hunk 19]
```diff
@@ -31,6 +31,143 @@ frequency: 2
 	</Resource>
 </Resources>
 
+## Two Pointers
+
+The two pointers method iterates two pointers across an array, to track the start and end
+of an interval. It can also be used to track two values in order to satisfy a specific condition.
```
Comment (andrewwangva) [inline content/3_Silver/2P.mdx:37]: still sorta unclear. To a new reader with no knowledge of 2p, "used to trace two values in order to satisfy a specific condition" is simply too broad and isn't very helpful.

File: content/3_Silver/2P.mdx [hunk 20]
```diff
@@ -31,6 +31,143 @@ frequency: 2
...
+
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.
+
+Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segment $\texttt{left} \dots \texttt{right}-1$, we can use two-pointers (or, alternatively, [binary search](/silver/binary-search)).
+
+For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
+
+We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
+
+We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$, the total time of current segment, and increment $\texttt{left}$ by 1.
```
Comment (andrewwangva) [inline content/3_Silver/2P.mdx:53]: Hard to follow. Instead focus on the continuous movement of the left pointer and at each moment the calculation of max books read.

> We can then remove the $\texttt{left}$-indexed book from $\texttt{cur}$, the total time of current segment, and increment $\texttt{left}$ by 1.

File: content/3_Silver/2P.mdx [hunk 21]
```diff
@@ -31,6 +31,143 @@ frequency: 2
 	</Resource>
 </Resources>
 
+## Two Pointers
+
+The two pointers method iterates two pointers across an array, to track the start and end
+of an interval. It can also be used to track two values in an array as show in CPH's 2SUM solution.
```
Comment (maggie-j-liu) [inline content/3_Silver/2P.mdx:37]:
```suggestion
of an interval. It can also be used to track two values in an array as shown in CPH's 2SUM solution. 
```

File: content/3_Silver/2P.mdx [hunk 22]
```diff
@@ -31,6 +31,143 @@ frequency: 2
...
+## Two Pointers
+
+The two pointers method iterates two pointers across an array, to track the start and end
+of an interval. It can also be used to track two values in an array as show in CPH's 2SUM solution. 
+
+<FocusProblem problem="book" />
+
+## Solution - Books
+
+We want to find the longest continuous segment of books that can be read within $t$ minutes.
```
Comment (maggie-j-liu) [inline content/3_Silver/2P.mdx:43]: is contiguous a better word? idk

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: content/3_Silver/2P.mdx (hunk 22 above)
Comment (HarryWangATX) [reply content/3_Silver/2P.mdx:43]: sure.

File: content/3_Silver/2P.mdx [hunk 23]
```diff
@@ -31,6 +31,143 @@ frequency: 2
...
+## Two Pointers
+
+The two pointers method iterates two pointers across an array, to track the start and end
+of an interval. It can also be used to track two values in an array as shown in CPH's 2SUM solution. 
+
+<FocusProblem problem="book" />
+
+## Solution - Books
+
+We want to find the longest continuous segment of books that can be read within $t$ minutes.
```
Comment (HarryWangATX) [inline content/3_Silver/2P.mdx:43]:
```suggestion
We want to find the longest contiguous segment of books that can be read within $t$ minutes. 
```

File: content/3_Silver/2P.mdx [hunk 24]
```diff
@@ -31,6 +31,143 @@ frequency: 2
 	</Resource>
 </Resources>
 
+## Two Pointers
+
+The two pointers method iterates two pointers across an array, to track the start and end
+of an interval. It can also be used to track two values in an array as shown in CPH's 2SUM solution.
```
Comment (andrewwangva) [inline content/3_Silver/2P.mdx:37]: maybe just link cph again?

File: content/3_Silver/2P.mdx [hunk 25]
```diff
@@ -31,6 +31,143 @@ frequency: 2
...
+
+## Solution - Books
+
+We want to find the longest contiguous segment of books that can be read within $t$ minutes. 
+
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively.
+
+Since being able to read segment $\texttt{left} \dots \texttt{right}$ implies we can read segment $\texttt{left} \dots \texttt{right}-1$, we can use two-pointers (or, alternatively, [binary search](/silver/binary-search)).
+
+For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
```
Comment (andrewwangva) [inline content/3_Silver/2P.mdx:49]: > For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
Kind of misleading - you're not really increasing unit it exceeds. It's more like increasing until it reaches max.

File: content/3_Silver/2P.mdx [hunk 26]
```diff
@@ -31,6 +31,143 @@ frequency: 2
...
+
+We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
+
+We can then increment $\texttt{left}$ by 1.
```
Comment (andrewwangva) [inline content/3_Silver/2P.mdx:53]: maybe say continually increment one to clarify?

File: content/3_Silver/2P.mdx (hunk 25 above)
Comment (HarryWangATX) [reply content/3_Silver/2P.mdx:49]: wdym by the max? im increasing it until it exceeds t right?

File: content/3_Silver/2P.mdx [hunk 27]
```diff
@@ -31,6 +31,141 @@ frequency: 2
...
+
+For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
+
+We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
+
+We can then continually increment $\texttt{left}$ by 1.
+
+**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right; therefore, since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**
+
+<Info title="Pro-tip">
```
Comment (maggie-j-liu) [inline content/3_Silver/2P.mdx:55]:
```suggestion
<Info title="Pro Tip">
```
to be consistent

File: content/3_Silver/2P.mdx [hunk 28]
```diff
@@ -31,6 +31,141 @@ frequency: 2
...
+You can visualize these pointers as maintaining a sliding window of books for this problem.
+
+</Info>
+
+### Implementation
+
+<LanguageSection>
+
+<CPPSection>
+
```
Comment (maggie-j-liu) [inline content/3_Silver/2P.mdx:66]:
```suggestion
```

File: content/3_Silver/2P.mdx [hunk 29]
```diff
@@ -31,6 +31,141 @@ frequency: 2
...
+#define pb push_back
+#define all(x) begin(x), end(x)
+#define sz(x) (int)(x).size()
+
+using pi = pair<int,int>;
+#define f first
+#define s second
+#define mp make_pair
+
+void setIO(string name = "") {
```
Comment (maggie-j-liu) [inline content/3_Silver/2P.mdx:84]: there's a codesnip for all this; wonder why we never use it?? it's `CPP Short Template`

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Review by maggie-j-liu — APPROVED** [review]

File: content/3_Silver/2P.mdx [hunk 30]
```diff
@@ -31,6 +31,118 @@ frequency: 2
...
+
+To accomplish this, we can define $\texttt{left}$ and $\texttt{right}$ as the start and end of the segment respectively and use two pointers to track the interval.
+
+For every $\texttt{left}$ index, we need to increase our $\texttt{right}$ pointer until the total time for that segment exceeds $t$.
+
+We will keep track of the max segment size with $\texttt{ans}$, and take the maximum $\texttt{right} - \texttt{left}$ value.
+
+We can then continually increment $\texttt{left}$ by 1.
+
+**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right; therefore, since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**
```
Comment (ar492) [inline content/3_Silver/2P.mdx:53]: I would have said something like "At each timestep, exactly one pointer is incremented. Since each pointer can be incremented at most N times, it takes at most 2N time, resulting in a time complexity of $\mathcal{O}(N)$.
```suggestion
**The $\texttt{right}$ pointer does not restart after an iteration, it simply keeps moving right; therefore, since both pointers will move at most $N$ times, the overall time complexity is $\mathcal{O}(N)$.**
```

Comment (ar492) [reply content/3_Silver/2P.mdx:53]: but what you said is fine as well

**Review by ar492 — APPROVED** [review]

**Review by frodakcin — APPROVED** [review]

File: content/3_Silver/2P.mdx (hunk 25 above)
Comment (andrewwangva) [reply content/3_Silver/2P.mdx:47]: well it isn't really _exceeding_ t right - it's more like increasing until it hits the maximum that's less than t.

Comment (HarryWangATX) [reply content/3_Silver/2P.mdx:47]: > well it isn't really _exceeding_ t right - it's more like increasing until it hits the maximum that's less than t.

@andrewwangva oh i c what u mean, but for each time step, we are increasing right until it exceeds t tho.

Comment (HarryWangATX) [reply content/3_Silver/2P.mdx:47]: oh wait i misread ur msg, ill make it clearer

Comment (HarryWangATX) [reply content/3_Silver/2P.mdx]: okay fixed.
